ingress/ingress.go

```diff
@@ -0,0 +1,150 @@
+)
+
+const (
+	ingressIndexName = "byIngress"
+)
+
+type IngressController struct {
+	kubeclientset    kubernetes.Interface
+	rolloutsIndexer  cache.Indexer
+	rolloutSynced    cache.InformerSynced
+	ingressLister    extentionslisters.IngressLister
+	ingressSynced    cache.InformerSynced
+	rolloutWorkqueue workqueue.RateLimitingInterface
+	ingressWorkqueue workqueue.RateLimitingInterface
+	resyncPeriod     time.Duration
+
+	metricServer   *metrics.MetricsServer
+	enqueueRollout func(obj interface{})
+}
+
+// NewIngressController returns a new ingress controller
+func NewIngressController(
+	kubeclientset kubernetes.Interface,
+	ingressInformer extensionsinformers.IngressInformer,
+	rolloutsInformer informers.RolloutInformer,
+	resyncPeriod time.Duration,
+	rolloutWorkQueue workqueue.RateLimitingInterface,
+	ingressWorkQueue workqueue.RateLimitingInterface,
+	metricServer *metrics.MetricsServer) *IngressController {
+
+	controller := &IngressController{
+		kubeclientset:   kubeclientset,
+		rolloutsIndexer: rolloutsInformer.Informer().GetIndexer(),
+		rolloutSynced:   rolloutsInformer.Informer().HasSynced,
+		ingressLister:   ingressInformer.Lister(),
+		ingressSynced:   ingressInformer.Informer().HasSynced,
+
+		rolloutWorkqueue: rolloutWorkQueue,
+		ingressWorkqueue: ingressWorkQueue,
+		resyncPeriod:     resyncPeriod,
+		metricServer:     metricServer,
+	}
+
+	util.CheckErr(rolloutsInformer.Informer().AddIndexers(cache.Indexers{
+		ingressIndexName: func(obj interface{}) (strings []string, e error) {
+			if rollout, ok := obj.(*v1alpha1.Rollout); ok {
+				return ingressutil.GetRolloutIngressKeys(rollout), nil
+			}
+			return []string{}, nil
+		},
+	}))
+
+	ingressInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			controllerutil.Enqueue(obj, ingressWorkQueue)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			controllerutil.Enqueue(newObj, ingressWorkQueue)
+		},
+		DeleteFunc: func(obj interface{}) {
+			controllerutil.Enqueue(obj, ingressWorkQueue)
+		},
+	})
+	controller.enqueueRollout = func(obj interface{}) {
+		controllerutil.EnqueueRateLimited(obj, rolloutWorkQueue)
+	}
+
+	return controller
+}
+
+func (c *IngressController) Run(threadiness int, stopCh <-chan struct{}) error {
+	log.Info("Starting Ingress workers")
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(func() {
+			controllerutil.RunWorker(c.ingressWorkqueue, logutil.ServiceKey, c.syncIngress, c.metricServer)
+		}, time.Second, stopCh)
+	}
+
+	log.Info("Started Ingress workers")
+	<-stopCh
+	log.Info("Shutting down Ingress workers")
+
+	return nil
+}
+
+func (c *IngressController) syncIngress(key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	_, err = c.ingressLister.Ingresses(namespace).Get(name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			if !strings.HasSuffix(name, "-canary") {
```

Comment:
this `-canary` suffix smells like a user-facing thing.  Is it documented somewhere?

Reply:
I agree. When we decided the controller should make the canaryIngress (rather than have the user supply it - see https://github.com/argoproj/argo-rollouts/issues/401#issuecomment-583649829) I needed to come up with some convention on what to name the canary copy. I opted for the `-canary` suffix, but this thing feels unclean to me.
What I am trying to do in this code is:
If something external is deleting the referenced stableIngress, do nothing (it'll just yield errors). If something external is deleting the canaryIngress, reconcile again to ensure it is recreated.
I would have preferred to introduce a custom annotation to check whether it's the stable or the canary ingress, but I don't have access to the full Ingress resource in this function, only namespace and name.

Reply:
I would be afraid that someone on the other side of the cluster is putting `-canary` on their name without knowing about the behavior on this side.  Is there any way this could affect something like that unintentionally?

Reply:
It's a valid concern, but I think the behavior of cloning the reference ingress it with`-canary` can be documented away. The ingress clone name needs to be deterministic (since the controller needs to handle handle retries of create properly in the event reconciliation fails), so randomly generated names using `generateName` is not an option. 

> I would be afraid that someone on the other side of the cluster is putting -canary on their name without knowing about the behavior on this side

To protect ourselves from operating on an ingress which was not created by the rollout, we'll need to verify the ownership reference of the ingress clone, matches the UID of the rollout. If it doesn't, it means it was created by someone else, and the rollout should have an error condition

Reply:
I'll add a check to validate the ownerRef if the canaryIngress already exists and error if it doesn't match expected.